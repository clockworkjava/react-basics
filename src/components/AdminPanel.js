import React from 'react';
import LoginPanel from './LoginPanel';
import AddBookForm from './AddBookForm';
import AdminBookListing from './AdminBookListing'
import {fbase} from '../fbase';

class AdminPanel extends React.Component {

    constructor() {
        super();
        this.state = {
            loggedIn : false,
            editMode : false,
            bookToEdit : {} 
        };
    };

    changeLoggedIn = (newValue) => this.setState({loggedIn: newValue}) 

    addNewBook = (book) => this.setState({
        books : [...this.state.books, book],
        editMode : false,
        bookToEdit : {}
    })

    componentDidMount() {
        this.ref = fbase.syncState('bookstore/books',{
            context: this,
            state: 'books'
        });
    }

    componentWillUnmount() {
       fbase.removeBinding(this.ref);
    }

    removeFromInventory = (title) => {
        this.setState({
            books : this.state.books.filter( book => title!==book.name )
        })
    }

    sendBookToEdit = (bookToEdit) => {
        this.setState({
            editMode: true,
            bookToEdit: bookToEdit
        });
    }

    editBook = (oldBookTitle,bookAfterEdit) => {
        
        const newBooks = this.state.books.filter( book => oldBookTitle!==book.name );

        this.setState({
            books : [...newBooks, bookAfterEdit],
            editMode : false,
            bookToEdit : {}
        })
    }
    render() {

        return (
            <div>
                {!this.state.loggedIn &&
                    <LoginPanel changeLoggedIn={this.changeLoggedIn}/>
                }
                {this.state.loggedIn && 
                    <React.Fragment>
                        <AddBookForm 
                            addNewBook = {this.addNewBook}    
                            editMode = {this.state.editMode}
                            book = {this.state.bookToEdit}
                            editBook = {this.editBook}
                        />
                        <AdminBookListing 
                            books = {this.state.books}
                            removeFromInventory = {this.removeFromInventory}
                            sendBookToEdit = {this.sendBookToEdit}
                        />
                    </React.Fragment>
                }
            </div>
        )}
}

export default AdminPanel;