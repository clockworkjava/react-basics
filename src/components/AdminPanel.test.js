import React from 'react';
import ReactDOM from 'react-dom';
import AdminPanel from './AdminPanel';

import {configure, shallow} from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';

configure({adapter: new Adapter()});

describe('AdminPanel tests', () => {

    it('AdminPanel renders without a problem', () => {
        const div = document.createElement('div');
        ReactDOM.render(<AdminPanel />, div);
        ReactDOM.unmountComponentAtNode(div);
    })

    it('Snapshot matches', () => {
        const wrapper = shallow(<AdminPanel/>);
        expect(wrapper).toMatchSnapshot();
    })

})

