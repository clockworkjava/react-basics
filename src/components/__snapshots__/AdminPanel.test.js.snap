// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AdminPanel tests Snapshot matches 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AdminPanel />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <form
        onSubmit={[Function]}
>
        <div
                className="form-group"
        >
                <input
                        className="form-control"
                        id="name"
                        name="name"
                        onChange={[Function]}
                        placeholder="Book name"
                        type="text"
                        value=""
                />
        </div>
        <div
                className="form-group"
        >
                <input
                        className="form-control"
                        id="author"
                        name="author"
                        onChange={[Function]}
                        placeholder="Book author"
                        type="text"
                        value=""
                />
        </div>
        <div
                className="form-group"
        >
                <textarea
                        className="form-control"
                        id="description"
                        name="description"
                        onChange={[Function]}
                        placeholder="Book description"
                        value=""
                />
        </div>
        <div
                className="form-group"
        >
                <input
                        className="form-check-input"
                        id="onStock"
                        name="onStock"
                        onChange={[Function]}
                        type="checkbox"
                        value={true}
                />
                <label
                        className="form-check-label"
                        htmlFor="onStock"
                >
                        On stock
                </label>
        </div>
        <div
                className="form-group"
        >
                <input
                        className="form-control"
                        id="image"
                        name="image"
                        onChange={[Function]}
                        placeholder="Book image"
                        type="text"
                        value=""
                />
        </div>
        <button
                className="btn btn-primary"
                type="submit"
        >
                Add
        </button>
</form>,
      "className": "adminPanel col-xs-4",
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="form-group"
>
            <input
                        className="form-control"
                        id="name"
                        name="name"
                        onChange={[Function]}
                        placeholder="Book name"
                        type="text"
                        value=""
            />
</div>,
          <div
            className="form-group"
>
            <input
                        className="form-control"
                        id="author"
                        name="author"
                        onChange={[Function]}
                        placeholder="Book author"
                        type="text"
                        value=""
            />
</div>,
          <div
            className="form-group"
>
            <textarea
                        className="form-control"
                        id="description"
                        name="description"
                        onChange={[Function]}
                        placeholder="Book description"
                        value=""
            />
</div>,
          <div
            className="form-group"
>
            <input
                        className="form-check-input"
                        id="onStock"
                        name="onStock"
                        onChange={[Function]}
                        type="checkbox"
                        value={true}
            />
            <label
                        className="form-check-label"
                        htmlFor="onStock"
            >
                        On stock
            </label>
</div>,
          <div
            className="form-group"
>
            <input
                        className="form-control"
                        id="image"
                        name="image"
                        onChange={[Function]}
                        placeholder="Book image"
                        type="text"
                        value=""
            />
</div>,
          <button
            className="btn btn-primary"
            type="submit"
>
            Add
</button>,
        ],
        "onSubmit": [Function],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <input
              className="form-control"
              id="name"
              name="name"
              onChange={[Function]}
              placeholder="Book name"
              type="text"
              value=""
/>,
            "className": "form-group",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form-control",
              "id": "name",
              "name": "name",
              "onChange": [Function],
              "placeholder": "Book name",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <input
              className="form-control"
              id="author"
              name="author"
              onChange={[Function]}
              placeholder="Book author"
              type="text"
              value=""
/>,
            "className": "form-group",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form-control",
              "id": "author",
              "name": "author",
              "onChange": [Function],
              "placeholder": "Book author",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <textarea
              className="form-control"
              id="description"
              name="description"
              onChange={[Function]}
              placeholder="Book description"
              value=""
/>,
            "className": "form-group",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form-control",
              "id": "description",
              "name": "description",
              "onChange": [Function],
              "placeholder": "Book description",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "textarea",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <input
                className="form-check-input"
                id="onStock"
                name="onStock"
                onChange={[Function]}
                type="checkbox"
                value={true}
/>,
              <label
                className="form-check-label"
                htmlFor="onStock"
>
                On stock
</label>,
            ],
            "className": "form-group",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form-check-input",
                "id": "onStock",
                "name": "onStock",
                "onChange": [Function],
                "type": "checkbox",
                "value": true,
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "On stock",
                "className": "form-check-label",
                "htmlFor": "onStock",
              },
              "ref": null,
              "rendered": "On stock",
              "type": "label",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": <input
              className="form-control"
              id="image"
              name="image"
              onChange={[Function]}
              placeholder="Book image"
              type="text"
              value=""
/>,
            "className": "form-group",
          },
          "ref": null,
          "rendered": Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "form-control",
              "id": "image",
              "name": "image",
              "onChange": [Function],
              "placeholder": "Book image",
              "type": "text",
              "value": "",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "Add",
            "className": "btn btn-primary",
            "type": "submit",
          },
          "ref": null,
          "rendered": "Add",
          "type": "button",
        },
      ],
      "type": "form",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <form
          onSubmit={[Function]}
>
          <div
                    className="form-group"
          >
                    <input
                              className="form-control"
                              id="name"
                              name="name"
                              onChange={[Function]}
                              placeholder="Book name"
                              type="text"
                              value=""
                    />
          </div>
          <div
                    className="form-group"
          >
                    <input
                              className="form-control"
                              id="author"
                              name="author"
                              onChange={[Function]}
                              placeholder="Book author"
                              type="text"
                              value=""
                    />
          </div>
          <div
                    className="form-group"
          >
                    <textarea
                              className="form-control"
                              id="description"
                              name="description"
                              onChange={[Function]}
                              placeholder="Book description"
                              value=""
                    />
          </div>
          <div
                    className="form-group"
          >
                    <input
                              className="form-check-input"
                              id="onStock"
                              name="onStock"
                              onChange={[Function]}
                              type="checkbox"
                              value={true}
                    />
                    <label
                              className="form-check-label"
                              htmlFor="onStock"
                    >
                              On stock
                    </label>
          </div>
          <div
                    className="form-group"
          >
                    <input
                              className="form-control"
                              id="image"
                              name="image"
                              onChange={[Function]}
                              placeholder="Book image"
                              type="text"
                              value=""
                    />
          </div>
          <button
                    className="btn btn-primary"
                    type="submit"
          >
                    Add
          </button>
</form>,
        "className": "adminPanel col-xs-4",
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="form-group"
>
              <input
                            className="form-control"
                            id="name"
                            name="name"
                            onChange={[Function]}
                            placeholder="Book name"
                            type="text"
                            value=""
              />
</div>,
            <div
              className="form-group"
>
              <input
                            className="form-control"
                            id="author"
                            name="author"
                            onChange={[Function]}
                            placeholder="Book author"
                            type="text"
                            value=""
              />
</div>,
            <div
              className="form-group"
>
              <textarea
                            className="form-control"
                            id="description"
                            name="description"
                            onChange={[Function]}
                            placeholder="Book description"
                            value=""
              />
</div>,
            <div
              className="form-group"
>
              <input
                            className="form-check-input"
                            id="onStock"
                            name="onStock"
                            onChange={[Function]}
                            type="checkbox"
                            value={true}
              />
              <label
                            className="form-check-label"
                            htmlFor="onStock"
              >
                            On stock
              </label>
</div>,
            <div
              className="form-group"
>
              <input
                            className="form-control"
                            id="image"
                            name="image"
                            onChange={[Function]}
                            placeholder="Book image"
                            type="text"
                            value=""
              />
</div>,
            <button
              className="btn btn-primary"
              type="submit"
>
              Add
</button>,
          ],
          "onSubmit": [Function],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <input
                className="form-control"
                id="name"
                name="name"
                onChange={[Function]}
                placeholder="Book name"
                type="text"
                value=""
/>,
              "className": "form-group",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form-control",
                "id": "name",
                "name": "name",
                "onChange": [Function],
                "placeholder": "Book name",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <input
                className="form-control"
                id="author"
                name="author"
                onChange={[Function]}
                placeholder="Book author"
                type="text"
                value=""
/>,
              "className": "form-group",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form-control",
                "id": "author",
                "name": "author",
                "onChange": [Function],
                "placeholder": "Book author",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <textarea
                className="form-control"
                id="description"
                name="description"
                onChange={[Function]}
                placeholder="Book description"
                value=""
/>,
              "className": "form-group",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form-control",
                "id": "description",
                "name": "description",
                "onChange": [Function],
                "placeholder": "Book description",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "textarea",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <input
                  className="form-check-input"
                  id="onStock"
                  name="onStock"
                  onChange={[Function]}
                  type="checkbox"
                  value={true}
/>,
                <label
                  className="form-check-label"
                  htmlFor="onStock"
>
                  On stock
</label>,
              ],
              "className": "form-group",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "className": "form-check-input",
                  "id": "onStock",
                  "name": "onStock",
                  "onChange": [Function],
                  "type": "checkbox",
                  "value": true,
                },
                "ref": null,
                "rendered": null,
                "type": "input",
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "host",
                "props": Object {
                  "children": "On stock",
                  "className": "form-check-label",
                  "htmlFor": "onStock",
                },
                "ref": null,
                "rendered": "On stock",
                "type": "label",
              },
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": <input
                className="form-control"
                id="image"
                name="image"
                onChange={[Function]}
                placeholder="Book image"
                type="text"
                value=""
/>,
              "className": "form-group",
            },
            "ref": null,
            "rendered": Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "form-control",
                "id": "image",
                "name": "image",
                "onChange": [Function],
                "placeholder": "Book image",
                "type": "text",
                "value": "",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Add",
              "className": "btn btn-primary",
              "type": "submit",
            },
            "ref": null,
            "rendered": "Add",
            "type": "button",
          },
        ],
        "type": "form",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
